include/nbl/builtin/hlsl/type_traits.hlsl

```diff
@@ -309,40 +350,130 @@ template <class T>
 struct is_aggregate : is_compound<T> {};
 
 template<class T>
-struct type_identity 
-{
-    using type = T;
-};
+struct rank : integral_constant<uint64_t, 0> { };
+
+template<class T, uint64_t N>
+struct rank<T[N]> : integral_constant<uint64_t, 1 + rank<T>::value> { };
+
+template<class T>
+struct rank<T[]> : integral_constant<uint64_t, 1 + rank<T>::value> { };
+
+template<class T, uint32_t I = 0> 
+struct extent : integral_constant<uint64_t, 0> {};
+
+template<class T, uint64_t N> 
+struct extent<T[N], 0> : integral_constant<uint64_t, N> {};
+
+template<class T, uint64_t N, uint32_t I> 
+struct extent<T[N], I> : integral_constant<uint64_t,extent<T, I - 1>::value> {};
+
+template<class T, uint32_t I> 
+struct extent<T[], I> : integral_constant<uint64_t,extent<T, I - 1>::value> {};
 
 template<bool B, class T = void>
 struct enable_if {};
  
 template<class T>
-struct enable_if<true, T> 
-{ 
-    using type = T; 
-};
+struct enable_if<true, T> : type_identity<T> {};
 
 // need this crutch because we can't make `#define typeid` work both on expression and types
 template<typename T>
 struct typeid_t;
 
+template<class T>
+struct alignment_of;
+
+
+// reference stuff needed for semantics 
+
+// not for "human consumption"
+// dxc doesnt actually allow variables to be references
+// every reference is treated as having 'restrict' qualifier
+// https://godbolt.org/z/dsj99fY96
 namespace impl
 {
 
-template<uint32_t encoded_typeid>
-struct decltype_t;
+template<class>
+struct make_void { using type = void; };
```

Comment:
can be outside of `impl`, also list `void_t`  in the comments above
https://github.com/Devsh-Graphics-Programming/Nabla/pull/546#discussion_r1346115889
tell us to use `make_void` instead, and point out that unfortunately `make_void` is not variadic